### PR TITLE
Micronaut Support for Resilience4j

### PR DESCRIPTION
**Pull request by pollend**
there is a task in the micronaut repo for adding support micronaut support to resilience4j: https://github.com/micronaut-projects/micronaut-core/issues/965

This is a bit of a mess but I'll clean this up as I get closer to a final implementation. 

here is some wip work that I've been working on to add support. Some of the methods in the factory classes are pretty close to their spring counterparts. not sure If I will have to tweak them but the implementation is pretty close between the two. I'm a little stuck with building out the interceptor methods for micronaut-aop and I'll have to put together some tests to verify If I implemented this correctly.

### CODE REVIEW

**Comment by pollend** [comment]
I kind of realize this is somewhat verbose. I was thinking about wrapping some of these handlers to reduce the code but I'm entirely sure if it is necessary.

**Comment by RobWin** [comment]
Thanks.
Can you tell us when the PR is ready to be revied?

**Comment by pollend** [comment]
@RobWin I think the main part of the PR is done. I just need to work on adding tests to verify the code.

**Comment by pollend** [comment]
kind of strange I can't work out why the tests seem to fail on the CI but work correctly on my local machine. I'm trying a couple things and I can't quite work out what is causing this behavior.

File: resilience4j-micronaut/src/test/groovy/io/github/resilience4j/micronaut/retry/RetySpec.groovy [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+import java.util.concurrent.CompletableFuture
+
+@MicronautTest
+@Property(name = "resilience4j.retry.enabled", value = "true")
+@Property(name = "resilience4j.retry.configs.default.maxRetryAttempts", value = "3")
+@Property(name = "resilience4j.retry.configs.default.waitDuration", value = "PT1S")
+@Property(name = "resilience4j.retry.configs.default.retryExceptions", value = "java.io.IOException")
+@Property(name = "resilience4j.retry.instances.backendA.baseConfig", value = "default")
+@Property(name = "resilience4j.retry.instances.backendA.maxRetryAttempts", value = "1")
+class RetySpec extends Specification{
```
Comment (RobWin) [inline resilience4j-micronaut/src/test/groovy/io/github/resilience4j/micronaut/retry/RetySpec.groovy:29]: Please rename to RetrySpec

File: resilience4j-micronaut/src/main/java/io/github/resilience4j/bulkhead/BulkheadSpecificationInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+
+/**
+ * A {@link MethodInterceptor} that intercepts all method calls which are annotated with a {@link io.github.resilience4j.annotation.Bulkhead}
+ * annotation.
+ **/
+@Singleton
+@Requires(classes = BulkheadRegistry.class)
+public class BulkheadSpecificationInterceptor implements MethodInterceptor<Object,Object> {
```
Comment (RobWin) [inline resilience4j-micronaut/src/main/java/io/github/resilience4j/bulkhead/BulkheadSpecificationInterceptor.java:51]: Why not only `BulkheadInterceptor`

File: resilience4j-micronaut/src/main/java/io/github/resilience4j/bulkhead/BulkheadSpecificationInterceptor.java [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+
+        Optional<AnnotationValue<io.github.resilience4j.annotation.Bulkhead>> opt = context.findAnnotation(io.github.resilience4j.annotation.Bulkhead.class);
+        if (!opt.isPresent()) {
+            return context.proceed();
+        }
+
+        ExecutableMethod executableMethod = context.getExecutableMethod();
+        final String name = executableMethod.stringValue(io.github.resilience4j.annotation.Bulkhead.class).orElse("default");
+
+        Bulkhead bulkhead = this.bulkheadRegistry.bulkhead(name);
```
Comment (RobWin) [inline resilience4j-micronaut/src/main/java/io/github/resilience4j/bulkhead/BulkheadSpecificationInterceptor.java:102]: We must check the type of Bulkhead `Bulkhead.Type.THREADPOOL` and create a ThreadPoolBulkhead if requested.

File: resilience4j-micronaut/src/main/java/io/github/resilience4j/ratelimiter/RateLimiterSpecificationInterceptor.java [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+
+import javax.inject.Singleton;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+
+@Singleton
+@Internal
+@Requires(classes = RateLimiterRegistry.class)
+public class RateLimiterSpecificationInterceptor implements MethodInterceptor<Object, Object> {
```
Comment (RobWin) [inline resilience4j-micronaut/src/main/java/io/github/resilience4j/ratelimiter/RateLimiterSpecificationInterceptor.java:46]: Can we rename it to RateLimiterInterceptor?

File: resilience4j-micronaut/src/main/java/io/github/resilience4j/retry/RetrySpecificationInterceptor.java [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+
+@Singleton
+@Internal
+@Requires(classes = RetryRegistry.class)
+public class RetrySpecificationInterceptor implements MethodInterceptor<Object,Object> {
```
Comment (RobWin) [inline resilience4j-micronaut/src/main/java/io/github/resilience4j/retry/RetrySpecificationInterceptor.java:48]: Can we rename it to `RetryInterceptor`

File: libraries.gradle [hunk 6]
```diff
@@ -32,6 +32,9 @@ ext {
     kotlinCoroutinesVersion = '1.3.2'
     springBootOpenFeignVersion = '2.1.2.RELEASE'
     blockhoundVersion = '1.0.1.RELEASE'
+    micronautVersion = '1.3.0.BUILD-SNAPSHOT'
```
[the remark this answers is not in the document]
Comment (pollend) [reply libraries.gradle:35]: I'm not sure actually. they use these versions with their other modules: 
- https://github.com/micronaut-projects/micronaut-kafka/blob/master/gradle.properties
- https://github.com/micronaut-projects/micronaut-sql/blob/master/gradle.properties
- https://github.com/micronaut-projects/micronaut-data/blob/master/gradle.properties

**Comment by pollend** [comment]
I cleaned up the code and remove a lot of the repetition in the code. I couldn't figure out how to handle ThreadPoolBulkhead for Flowable case?

File: libraries.gradle (hunk 6 above)
[the remark this answers is not in the document]
Comment (RobWin) [reply libraries.gradle:35]: We should always use the latest released version

**Comment by RobWin** [comment]
Yes, ThreadPoolBulkhead cannot be used for reactive return types. 
Please have a look at [BulkheadAspect](https://github.com/resilience4j/resilience4j/blob/37230a9c10ef9a3a872f406ac798e7b895a04502/resilience4j-spring/src/main/java/io/github/resilience4j/bulkhead/configure/BulkheadAspect.java#L243)
ThreadPoolBulkhead only supports CompletionStage as a return type, otherwise an exception is thrown.

**Comment by pollend** [comment]
Then I probably used it correctly.

File: libraries.gradle (hunk 6 above)
[the remark this answers is not in the document]
Comment (pollend) [reply libraries.gradle:35]: what version would that be? I think they're going to release 2.0 soon so you want to use that version?

[the remark this answers is not in the document]
Comment (RobWin) [reply libraries.gradle:35]: The latest release is 1.3.5, or?

[the remark this answers is not in the document]
Comment (pollend) [reply libraries.gradle:35]: I can bump it to 1.3.5, there is also the discussion in micronaut-core https://github.com/micronaut-projects/micronaut-core/issues/965 if you need help finding a version that would be suitable. there is also this issue with micronaut 2.0, https://github.com/micronaut-projects/micronaut-core/issues/3024

[the remark this answers is not in the document]
Comment (pollend) [reply libraries.gradle:35]: I tried it with 1.3.5 and the configuration does not seem to be written correctly.

```
Condition not satisfied:

defaultBulkhead.bulkheadConfig.maxWaitDuration.seconds == 10
|               |              |               |       |
|               |              PT0S            0       false
|               <io.github.resilience4j.bulkhead.BulkheadConfig@7fb7546f maxConcurrentCalls=25 maxWaitDuration=PT0S writableStackTraceEnabled=true>
Bulkhead 'default'

Condition not satisfied:

defaultBulkhead.bulkheadConfig.maxWaitDuration.seconds == 10
|               |              |               |       |
|               |              PT0S            0       false
|               <io.github.resilience4j.bulkhead.BulkheadConfig@7fb7546f maxConcurrentCalls=25 maxWaitDuration=PT0S writableStackTraceEnabled=true>
Bulkhead 'default'

	at io.github.resilience4j.micronaut.bulkhead.BulkheadSpec.default configuration(BulkheadSpec.groovy:40)

```

**Comment by pollend** [comment]
Any clue why the azure pipeline seems be stuck and timing out when running resilience4j.reactor. 
```
> Task :resilience4j-reactor:test

io.github.resilience4j.reactor.retry.RetryOperatorTest > retryOnResultUsingMono FAILED
    java.lang.AssertionError at RetryOperatorTest.java:186

io.github.resilience4j.reactor.retry.RetryOperatorTest > retryOnResultFailAfterMaxAttemptsUsingFlux FAILED
    java.lang.AssertionError at RetryOperatorTest.java:267

io.github.resilience4j.reactor.retry.RetryOperatorTest > retryOnResultFailAfterMaxAttemptsUsingMono FAILED
    java.lang.AssertionError at RetryOperatorTest.java:209

io.github.resilience4j.reactor.retry.RetryOperatorTest > shouldFailWithExceptionFlux FAILED
    java.lang.AssertionError at RetryOperatorTest.java:224

io.github.resilience4j.reactor.retry.RetryOperatorTest > returnOnCompleteUsingMono FAILED
    java.lang.AssertionError at RetryOperatorTest.java:75

io.github.resilience4j.reactor.retry.RetryOperatorTest > returnOnErrorUsingMono FAILED
    java.lang.AssertionError at RetryOperatorTest.java:134

io.github.resilience4j.reactor.timelimiter.TimeLimiterOperatorTest > timeoutUsingFlux FAILED
    java.lang.AssertionError at TimeLimiterOperatorTest.java:132

io.github.resilience4j.reactor.timelimiter.TimeLimiterOperatorTest > timeoutUsingMono FAILED
    java.lang.AssertionError at TimeLimiterOperatorTest.java:69

io.github.resilience4j.reactor.timelimiter.TimeLimiterOperatorTest > timeoutNeverUsingFlux FAILED
    java.lang.AssertionError at TimeLimiterOperatorTest.java:147

io.github.resilience4j.reactor.timelimiter.TimeLimiterOperatorTest > timeoutNeverUsingMono FAILED
    java.lang.AssertionError at TimeLimiterOperatorTest.java:84
```

**Comment by RobWin** [comment]
Yes, please ignore Azure Pipelines. I've removed it from the master branch.

**Comment by pollend** [comment]
I know the test are lacking could it be improved at a later point or would it be worthwhile adding those all now.

**Comment by RobWin** [comment]
We try to achieve a test coverage of at least 80%. Please add some tests.

**Comment by pollend** [comment]
That should be the last thing I work out. can you review the codebase for any additional changes I would need to do?

**Comment by pollend** [comment]
I should work on that in the mean time.

**Comment by pollend** [comment]
so a lot of the repeated code was tweaked so could you review those changes and I should be working on adding a lot more test cases in the meantime.

**Comment by pollend** [comment]
So I'm using microanut AOP to intercept calls but is there any specific order I should expect in regards to handling bulkhead, circuitbreaker, bulkhead, and retry. the order is same across the board but I can change the index slightly so one is after the other and vice versa. I made sure to put it before @Fallback annotation that comes default with Micronaut.  if it errors out then it can be handled by that if someone chooses to use it.

**Comment by graemerocher** [comment]
@pollend @RobWin let me know if you need any help taking this forward

**Comment by RobWin** [comment]
@pollend The order should be similar to the default order used in Spring Boot ->  https://resilience4j.readme.io/docs/getting-started-3#aspect-order

**Comment by graemerocher** [comment]
note you may want to use https://github.com/micronaut-projects/micronaut-core/blob/1.3.x/aop/src/main/java/io/micronaut/aop/InterceptPhase.java#L43 as a starting point to define the order

**Comment by pollend** [comment]
Made some tweaks for the intercept order that should follow the document that was provided. Trying to figure out what else I should tweak in the PR.

**Comment by RobWin** [comment]
Can a user change the order of the interceptors via configuration?

**Comment by graemerocher** [comment]
Why would a user want to do that?

**Comment by RobWin** [comment]
In Spring Boot we allow the user to control the order of aspects so that the user can decide if you wants:

`Retry ( CircuitBreaker ( RateLimiter ( TimeLimiter ( Bulkhead ( Function ) ) ) ) )`

or 

`CircuitBreaker ( Retry ( RateLimiter ( TimeLimiter ( Bulkhead ( Function ) ) ) ) )`

or 

`RateLimiter( Retry ( CircuitBreaker ( TimeLimiter ( Bulkhead ( Function ) ) ) ) )`

You get the point ;)

**Comment by graemerocher** [comment]
hmmm and do users actually use the feature? Seems like it would complicate the code unnecessarily, but maybe I am missing something.

**Comment by RobWin** [comment]
People always use defaults, until they have special requirements. The functional style allows maximum flexibility. It would be nice if annotations + interceptors also allow it.

**Comment by pollend** [comment]
I saw this PR and it seems like the change that would allow for this special kind of ordering. 

https://github.com/resilience4j/resilience4j/pull/744

**Comment by pollend** [comment]
@RobWin when will you be available to review this?

**Comment by RobWin** [comment]
Hopefully soon ;)
I think we need to be able to have an early SNAPSHOT and update our micronaut-demo to test the features early.

**Comment by pollend** [comment]
@RobWin Anything I can help in regards to the micronaut-demo?

**Comment by RobWin** [comment]
We could create a micronaut branch and merge it into this branch instead of master. 
I will then publish a SNAPSHOT version which we can use to adapt https://github.com/resilience4j/resilience4j-micronaut-demo

What do you think?

**Comment by RobWin** [comment]
I created another PR for it 
https://github.com/resilience4j/resilience4j/pull/1036

**Comment by pollend** [comment]
@RobWin oh nice, I can start on that soonish.

**Comment by pollend** [comment]
@RobWin I have this branch I'm working on that uses the annotationProcessors to annotate the default annotations that comes with resiliance4j. I'm not quite sure what is wrong and I am wondering if this is worth the effort adding. It just removes having annotations repeated in two places of the library if lets say this is used with spring and Micronaut. 

https://github.com/pollend/resilience4j/tree/feature/micronaut-meta-inf

**Comment by graemerocher** [comment]
@pollend nice work!

**Comment by RobWin** [comment]
Thank you.
@graemerocher Have you seen such an exception before? 


```
Caused by: java.io.UncheckedIOException: java.util.zip.ZipException: ZipFile invalid LOC header (bad signature)
	at io.micronaut.core.io.service.SoftServiceLoader$ServiceLoaderIterator.hasNext(SoftServiceLoader.java:215)
```

**Comment by graemerocher** [comment]
I think that may be happening in gradle when you try to use an annotation processor defined in the same project from another subproject. Try removing https://github.com/pollend/resilience4j/blob/feature/micronaut-meta-inf/resilience4j-micronaut-annotation/build.gradle#L9

**Comment by pollend** [comment]
still not quite working, so I reworked the gradle file to look like this: 
```
repositories {
    maven { url "https://repo.grails.org/grails/core" }
    maven { url "https://oss.jfrog.org/oss-snapshot-local" }
}
dependencies {
    compileOnly "io.micronaut:micronaut-inject-java:$micronautVersion"
    compileOnly project(':resilience4j-annotations')
}
ext.moduleName = 'io.github.resilience4j.mapper'
```
https://gist.github.com/pollend/f20cd0831bbcee0bd8a53f8b70ab0bca

**Comment by graemerocher** [comment]
Yeah it is almost certainly something to do with using an annotation processor in the same project where it is defined but I am not sure, if you have found a workaround that works I would go for it. As you said it is likely that it works fine in an end user project.

**Comment by pollend** [comment]
If its not too bad then having it the annotation in two places is the current solution. @RobWin any status on the sample repo?

**Comment by graemerocher** [comment]
What are the steps to reproduce the Zip error?

**Comment by RobWin** [comment]
@pollend To be honest. I didn't have time to look at the Micronaut repo in more detail yet.
Too much work at work :(
But I will do it after my holiday.

**Comment by pollend** [comment]
Looks like it doesn't like the version of java i'm running. The other problem AnnotationMapper does not seem to work with the any of the interceptors. 

```
openjdk 11.0.7 2020-04-14
OpenJDK Runtime Environment (build 11.0.7+10-post-Ubuntu-3ubuntu1)
OpenJDK 64-Bit Server VM (build 11.0.7+10-post-Ubuntu-3ubuntu1, mixed mode, sharing)

```

**Comment by Romeh** [comment]
@pollend  right now resilience4j tested on Java 8 , we have another PR to upgrade to java 11 here https://github.com/resilience4j/resilience4j/pull/1089
can u please try again with Java 8 instead ?

**Comment by pollend** [comment]
So, I've been also working on a micronaut ignite library and I'm trying to do something similar with AOP and I'm finding that none of the intercepts get called. an intercept method is compiled onto the class when build and I'm finding the list of intercepts to be empty. This is because it can't find the Type when it looks up Around annotation because type is not a stereotype of Around. not sure this can be set from AnnotationMapper. Or there is some other way to achieve a similar affect. 

https://github.com/micronaut-projects/micronaut-core/blob/master/aop/src/main/java/io/micronaut/aop/chain/InterceptorChain.java#L283

**Comment by pollend** [comment]
@RobWin anything else that needs to be done? The demo should be in a good state for review when you get the chance.  

https://github.com/resilience4j/resilience4j-micronaut-demo/pull/2

**Comment by RobWin** [comment]
Sry too much work and PRs at the moment :/ 
I haven't forgot your PR.

**Comment by RobWin** [comment]
I create a new PR to merge and review the "micronaut" branch and close this one, ok?

**Comment by pollend** [comment]
sounds good, I'll be around.

**Pull request closed without merging**